REL-2232 fix nifs partitioner

tpolecat opened the pull request:
NIFS targets were not being partitioned correctly, so I fixed that and added some support for end-to-end testing of skeleton generation. See the included [`REL_2232_Test.scala`](https://github.com/tpolecat/ocs/blob/nifs-partitioner/bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/REL_2232_Test.scala) for an example. It is likely that the template expansion testing support will be further generalized when I update the Graces tests (after this is merged).

cquiroz commented:
Looks good :+1:

tpolecat commented:
Lots of new stuff, needs new review. Sorry.

cquiroz commented on bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Observation.scala:
```diff
@@ -11,6 +11,7 @@ object Observation {
   val condition:Lens[Observation, Option[Condition]] = Lens.lensu((a, b) => a.copy(condition = b), _.condition)
   val target:Lens[Observation, Option[Target]] = Lens.lensu((a, b) => a.copy(target = b), _.target)
   val meta:Lens[Observation, Option[ObservationMeta]] = Lens.lensu((a, b) => a.copy(meta = b), _.meta)
+  val time:Lens[Observation, Option[TimeAmount]] = Lens.lensu((a, b) => a.copy(time = b), _.time)
```
This is unrelated to this PR. What's the state of the art on generating (e.g. via Macros) lenses, I remember some discussions about that

tpolecat replied:
State of the art is a library by Julien Truffaut called [Monocle](https://github.com/julien-truffaut/Monocle). I would like to integrate it once we have upgraded to scalaz 7.1 (or 7.2, depending on when it happens).

cquiroz commented on bundle/edu.gemini.phase2.skeleton.servlet/skeleton.sh:
```diff
@@ -0,0 +1,106 @@
+#!/bin/bash
```
I thought `skeleton.sh` was present somewhere on the codebase (It may have been on the svn repo), but in any case great to have it here

cquiroz commented on bundle/edu.gemini.phase2.skeleton.servlet/src/main/scala/edu/gemini/phase2/skeleton/servlet/SkeletonServlet.scala:
```diff
@@ -104,6 +104,10 @@ final class SkeletonServlet(odb: IDBDatabaseService, templateFactory: TemplateFa
 
+  /** If `test` is set then the PDF file is optional and ignored if present. */
+  private def test(items: List[FileItem]): Boolean =
+    items.exists(_.getFieldName == "test")
```
Where does this field come from, is it int the proposal?

tpolecat replied:
I added it to the servlet so you can skip the PDF and attachment processing. Thanks for pointing this out; I had forgotten about it and need to write a test.

cquiroz commented on bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala:
```diff
@@ -0,0 +1,168 @@
+abstract class TemplateSpec(xmlName: String) { this: Specification =>
+
+  // Required when constructing phase 1 proposals
+  if (System.getProperty("edu.gemini.model.p1.schemaVersion") == null)
```
Hmm, I thought this was removed, and we were reading it from a class, but perhaps I'm talking about the semester

tpolecat replied:
There may be a better way to do this, but I was unable to construct a `Proposal` without it.

cquiroz commented on bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala:
```diff
@@ -0,0 +1,168 @@
+  def p1Obs(bp: BlueprintBase, st: Target): Observation =
+    Observation.empty execState {
+      for {
+        _ <- blueprint := Some(bp)
```
This is new to me, what`s the `:=`?

tpolecat replied:
`Lens[A,B]` can also be thought of as a pair of operations `get: State[A,B]` and `set: B => State[A, Unit]` and the second is what the `:=` does ... it just lets you do a bunch of lens operations as a program in `State`, which is sometimes easier to follow. Ed Kmett's [lecture](https://www.youtube.com/watch?v=efv0SQNde5Q) here explains this style of lens if you're interested.

cquiroz commented:
Great to have the template testing :+1: 
If I understand it correctly you can skip the pdf processing on the tests with these modifications, right?

tpolecat commented:
@cquiroz yes, the testing stuff at this point doesn't do any attachment processing; it's all in-memory and has no file IO at all.

tpolecat commented:
I removed the change to the skeleton servlet because it's not relevant here, and renamed the `testing` parameter to `preserveLibraryIds` which is more descriptive.
